llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/SplitReduction.cpp

```diff
@@ -0,0 +1,242 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.h"
+#include "iree-dialects/Dialect/LinalgExt/Transforms/Transforms.h"
+#include "iree-dialects/Dialect/LinalgExt/Transforms/Utils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "llvm/ADT/STLExtras.h"
+
+using namespace mlir;
+using namespace mlir::iree_compiler::IREE::LinalgExt;
+
+namespace {
+
+SmallVector<int64_t> getExpandedShape(ArrayRef<int64_t> inputShape,
+                                      int64_t splitReductionRatio,
+                                      int64_t splitDimParallel) {
+  SmallVector<int64_t> ans;
+  for (auto elem : llvm::enumerate(inputShape)) {
+    if (elem.index() == splitDimParallel) {
+      ans.push_back(splitReductionRatio);
+      ans.push_back(inputShape[splitDimParallel] / splitReductionRatio);
+    } else {
+      ans.push_back(elem.value());
+    }
+  }
+  return ans;
+}
+
+SmallVector<int64_t> getCollapsedShape(ArrayRef<int64_t> inputShape,
+                                       int64_t splitReductionRatio, int64_t k,
+                                       int64_t kDimOrig) {
+  SmallVector<int64_t> ans;
+  for (auto elem : inputShape) {
+    ans.push_back(elem);
+  }
+  ans[kDimOrig] = k * splitReductionRatio;
+  return ans;
+}
+
+} // namespace
+
+// Transforms an applicable standard single reduction TopkOp into a parallel
+// reduction TopkOp with a reduce step following.
+//
+// Handles parallel reductions in 2 phases: A "map" parallel phase and the a
+// single "reduce" reduction phase. The first phase expands the input tensor
+// shape by breaking the reduction dimension into multiple parallel reductions
+// (upping the rank of the input). Topk is run on these dimensions in parallel
+// The second phase collapses the parallel results into a single final reduce.
+// Topk is run again on the combined output to produce a final output.
+//
+// Currently only topk operations without input indices are supported.
+FailureOr<TopkOp> mlir::iree_compiler::IREE::LinalgExt::TopkOpSplitReduction::
+    returningMatchAndRewrite(iree_compiler::IREE::LinalgExt::TopkOp topkOp,
```

Comment:
This is not the expected form of `returningMatchAndRewrite`. I'd just implement the `matchAndRewrite` method directly. We can refactor later if we need to.

Reply:
Can you elaborate on why this is not the expected form? I'm following the form that all the other `OpRewritePattern`s in `Transforms.h`

Reply:
I don't know the details about it in `Transforms.h`. Maybe that's more for inheritance purpose. Let's follow the style from MLIR doc. We usually use `matchAndRewrite` directly.

https://mlir.llvm.org/docs/PatternRewriter/#match-and-rewrite-implementation

Reply:
This kind of interface is used upstream by some transformations in order to work with the transform dialect. It was suggested by Nicolas when writing the splitk pattern, it’s fine to skip if there are concerns. Not sure it will matter for some time.